src/DistributedTriangulations.jl

```diff
@@ -1,88 +1,58 @@
 
-function remove_ghost_cells(trian::Triangulation,part::Integer,gids::IndexSet)
+function filter_cells_when_needed(strategy::AssemblyStrategy, trian::Triangulation)
+  @abstractmethod
+end
+
+function filter_cells_when_needed(strategy::RowsComputedLocally, trian::Triangulation)
+  trian
+end
+
+function filter_cells_when_needed(strategy::OwnedCellsStrategy, trian::Triangulation)
+  remove_ghost_cells(trian,strategy.part,strategy.cell_gids)
+end
+
+function Gridap.Geometry.Triangulation(strategy::AssemblyStrategy,model::DiscreteModel,args...)
+  trian = Triangulation(model,args...)
+  filter_cells_when_needed(strategy,trian)
+end
+
+function Gridap.Geometry.BoundaryTriangulation(strategy::AssemblyStrategy,model::DiscreteModel,args...)
+  trian = BoundaryTriangulation(model,args...)
+  filter_cells_when_needed(strategy,trian)
+end
+
+function Gridap.Geometry.SkeletonTriangulation(strategy::AssemblyStrategy,model::DiscreteModel,args...)
+  trian = SkeletonTriangulation(model,args...)
+  filter_cells_when_needed(strategy,trian)
+end
+
+function remove_ghost_cells(trian::Triangulation, part::Integer, gids::IndexSet)
     tcell_to_mcell = get_cell_id(trian)
-    mcell_to_isowned = gids.lid_to_owner .== part
-    tcell_to_isowned = reindex(mcell_to_isowned,tcell_to_mcell)
-    ocell_to_tcell = findall(tcell_to_isowned)
-    TriangulationPortion(trian,ocell_to_tcell)
+    ocell_to_tcell =
+        findall((x) -> (gids.lid_to_owner[x] == part), tcell_to_mcell)
+    TriangulationPortion(trian, ocell_to_tcell)
+end
+
+function remove_ghost_cells(
+    trian::SkeletonTriangulation,
+    part::Integer,
+    gids::IndexSet,
+)
+    cell_id_left = get_cell_id(trian.left)
+    cell_id_right = get_cell_id(trian.right)
+    @assert length(cell_id_left) == length(cell_id_right)
+    facets_to_old_facets = eltype(cell_id_right)[]
```

Comment:
@amartinhuertas  I am not sure if this function is 100% type stable. Perhaps a function barrier for the loop will help. The `trian` object could be type in-stable (now or in the future) since it is a quite high-level object.

Reply:
Damm! I always forget about the issue of type instability. I will definitely add the function barrier. But I do not understand why do u say that `trian` could be type `instable` as it is declared to be of type `SkeletonTriangulation`. `trian.left` and `trian.right` types are also known. So I do not see whats wrong in this particular scenario.

Reply:
For my understanding, I ran the type inference algorithm (in the debugger), and as I expected, there is no particular type instability (now). Results below. Can you think of a future scenario in which there might be type instability? May be for a different implementation of `SkeletonTriangulation` ?

```
debug> InteractiveUtils.@code_warntype remove_ghost_cells(trian,strategy.part,strategy.cell_gids)
Variables
  #self#::Core.Compiler.Const(GridapDistributed.remove_ghost_cells, false)
  trian::Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)}
  part::Int64
  gids::GridapDistributed.IndexSet{Array{Int64,1},Array{Int64,1},Dict{Int64,Int32}}
  #87::GridapDistributed.var"#87#88"{Int64,GridapDistributed.IndexSet{Array{Int64,1},Array{Int64,1},Dict{Int64,Int32}}}
  tcell_to_mcell::Gridap.Arrays.IdentityVector{Int64}
  ocell_to_tcell::Array{Int64,1}

Body::Gridap.Geometry.TriangulationPortion{2,2,Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)}}
1 ─      (tcell_to_mcell = GridapDistributed.get_cell_id(trian))
│   %2 = GridapDistributed.:(var"#87#88")::Core.Compiler.Const(GridapDistributed.var"#87#88", false)
│   %3 = Core.typeof(part)::Core.Compiler.Const(Int64, false)
│   %4 = Core.typeof(gids)::Core.Compiler.Const(GridapDistributed.IndexSet{Array{Int64,1},Array{Int64,1},Dict{Int64,Int32}}, false)
│   %5 = Core.apply_type(%2, %3, %4)::Core.Compiler.Const(GridapDistributed.var"#87#88"{Int64,GridapDistributed.IndexSet{Array{Int64,1},Array{Int64,1},Dict{Int64,Int32}}}, false)
│        (#87 = %new(%5, part, gids))
│   %7 = #87::GridapDistributed.var"#87#88"{Int64,GridapDistributed.IndexSet{Array{Int64,1},Array{Int64,1},Dict{Int64,Int32}}}
│        (ocell_to_tcell = GridapDistributed.findall(%7, tcell_to_mcell))
│   %9 = GridapDistributed.TriangulationPortion(trian, ocell_to_tcell)::Gridap.Geometry.TriangulationPortion{2,2,Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)}}
└──      return %9

debug> InteractiveUtils.@code_warntype remove_ghost_cells(trian,strategy.part,strategy.cell_gids)
Variables
  #self#::Core.Compiler.Const(GridapDistributed.remove_ghost_cells, false)
  trian::Gridap.Geometry.SkeletonTriangulation{1,2,Gridap.Geometry.GenericBoundaryTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.UnstructuredGrid{1,2,Float64,false}},Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)},true,Int64}}
  part::Int64
  gids::GridapDistributed.IndexSet{Array{Int64,1},Array{Int64,1},Dict{Int64,Int32}}
  cell_id_left::Array{Int64,1}
  cell_id_right::Array{Int64,1}
  facets_to_old_facets::Array{Int64,1}
  @_8::Union{Nothing, Tuple{Int64,Int64}}
  i::Int64
  part_left::Int64
  part_right::Int64
  max_part_id::Int64

Body::Gridap.Geometry.SkeletonTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.GenericBoundaryTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.UnstructuredGrid{1,2,Float64,false}},Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)},true,Int64}}}
1 ─       Core.NewvarNode(:(facets_to_old_facets))
│         Core.NewvarNode(:(@_8))
│   %3  = Base.getproperty(trian, :left)::Gridap.Geometry.GenericBoundaryTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.UnstructuredGrid{1,2,Float64,false}},Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)},true,Int64}
│         (cell_id_left = GridapDistributed.get_cell_id(%3))
│   %5  = Base.getproperty(trian, :right)::Gridap.Geometry.GenericBoundaryTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.UnstructuredGrid{1,2,Float64,false}},Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)},true,Int64}
│         (cell_id_right = GridapDistributed.get_cell_id(%5))
│   %7  = GridapDistributed.length(cell_id_left)::Int64
│   %8  = GridapDistributed.length(cell_id_right)::Int64
│   %9  = (%7 == %8)::Bool
└──       goto #3 if not %9
2 ─       goto #4
3 ─ %12 = Base.AssertionError("length(cell_id_left) == length(cell_id_right)")::AssertionError
└──       Base.throw(%12)
4 ┄ %14 = GridapDistributed.eltype(cell_id_right)::Core.Compiler.Const(Int64, false)
│         (facets_to_old_facets = Base.getindex(%14))
│   %16 = GridapDistributed.length(cell_id_left)::Int64
│   %17 = (1:%16)::Core.Compiler.PartialStruct(UnitRange{Int64}, Any[Core.Compiler.Const(1, false), Int64])
│         (@_8 = Base.iterate(%17))
│   %19 = (@_8 === nothing)::Bool
│   %20 = Base.not_int(%19)::Bool
└──       goto #9 if not %20
5 ┄ %22 = @_8::Tuple{Int64,Int64}::Tuple{Int64,Int64}
│         (i = Core.getfield(%22, 1))
│   %24 = Core.getfield(%22, 2)::Int64
│   %25 = Base.getproperty(gids, :lid_to_owner)::Array{Int64,1}
│   %26 = Base.getindex(cell_id_left, i)::Int64
│         (part_left = Base.getindex(%25, %26))
│   %28 = Base.getproperty(gids, :lid_to_owner)::Array{Int64,1}
│   %29 = Base.getindex(cell_id_right, i)::Int64
│         (part_right = Base.getindex(%28, %29))
│         (max_part_id = GridapDistributed.max(part_left, part_right))
│   %32 = (max_part_id == part)::Bool
└──       goto #7 if not %32
6 ─       GridapDistributed.push!(facets_to_old_facets, i)
7 ┄       (@_8 = Base.iterate(%17, %24))
│   %36 = (@_8 === nothing)::Bool
│   %37 = Base.not_int(%36)::Bool
└──       goto #9 if not %37
8 ─       goto #5
9 ┄ %40 = GridapDistributed.TriangulationPortion(trian, facets_to_old_facets)::Gridap.Geometry.SkeletonTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.GenericBoundaryTriangulation{1,2,Gridap.Geometry.TriangulationPortion{1,2,Gridap.Geometry.UnstructuredGrid{1,2,Float64,false}},Gridap.Geometry.CartesianGrid{2,Float64,typeof(identity)},true,Int64}}}
└──       return %40
```

Reply:
`trian.left` or `trian.right` can potentially be a type in-stable specialization of Triangulation. For now it is not the case, but who knows in the future...

Reply:
>trian.left or trian.right can potentially be a type in-stable specialization of Triangulation. For now it is not the case, but who knows in the future...

Ok. This is what I was thinking. Thanks for confirming!

Reply:
> @amartinhuertas I am not sure if this function is 100% type stable. Perhaps a function barrier for the loop will help. The trian object could be type in-stable (now or in the future) since it is a quite high-level object.

Done in cbe9d3a22fe3937b95dbe10f4331e357c8db6355